pkg/apis/velero/v1/restore.go

```diff
@@ -80,6 +83,100 @@ type RestoreSpec struct {
 	// +optional
 	// +nullable
 	IncludeClusterResources *bool `json:"includeClusterResources,omitempty"`
+
+	// Hooks represent custom behaviors that should be executed during or post restore.
+	// +optional
+	Hooks RestoreHooks `json:"hooks,omitempty"`
+}
+
+// RestoreHooks contains custom behaviors that should be executed during or post restore.
+type RestoreHooks struct {
+	Resources []RestoreResourceHookSpec `json:"resources,omitempty"`
+}
+
+// RestoreResourceHookSpec defines one or more RestoreResrouceHooks that should be executed based on
+// the rules defined for namespaces, resources, and label selector.
+type RestoreResourceHookSpec struct {
+	// Name is the name of this hook.
+	Name string `json:"name"`
+
+	// IncludedNamespaces specifies the namespaces to which this hook spec applies. If empty, it applies
+	// to all namespaces.
+	// +optional
+	// +nullable
+	IncludedNamespaces []string `json:"includedNamespaces,omitempty"`
+
+	// ExcludedNamespaces specifies the namespaces to which this hook spec does not apply.
+	// +optional
+	// +nullable
+	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`
+
+	// IncludedResources specifies the resources to which this hook spec applies. If empty, it applies
+	// to all resources.
+	// +optional
+	// +nullable
+	IncludedResources []string `json:"includedResources,omitempty"`
+
+	// ExcludedResources specifies the resources to which this hook spec does not apply.
+	// +optional
+	// +nullable
+	ExcludedResources []string `json:"excludedResources,omitempty"`
+
+	// LabelSelector, if specified, filters the resources to which this hook spec applies.
+	// +optional
+	// +nullable
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// PostHooks is a list of RestoreResourceHooks to execute during and after restoring a resource.
+	// +optional
+	PostHooks []RestoreResourceHook `json:"post,omitempty"`
+}
+
+// RestoreResourceHook defines a restore hook for a resource.
+type RestoreResourceHook struct {
+	// Exec defines an exec restore hook.
+	Exec *ExecRestoreHook `json:"exec"`
+
+	// Init defines an init restore hook.
+	Init *InitRestoreHook `json:"init"`
+}
+
+// ExecRestoreHook is a hook that uses pod exec API to execute a command inside a container in a pod
+type ExecRestoreHook struct {
+	// Container is the container in the pod where the command should be executed. If not specified,
+	// the pod's first container is used.
+	// +optional
+	Container string `json:"container,omitempty"`
+
+	// Command is the command and arguments to execute from within a container after a pod has been restored.
+	// +kubebuilder:validation:MinItems=1
```

Comment:
Is this mixing kubebuilder and non-kubebuilder generation in the same type, since this is eventually included in `Restore`?

Reply:
Also - since this is set to at least 1 item in the field, how does that interact with `ExecRestoreHooks` being optional on the `RestoreSpec`? Does it mean that the validation will happen if the `ExecRestoreHook` is non-empty?

Reply:
The `kubebuilder` annotations are all copied from the `BackupSpec`.
But to answer your question, yes the Validation will happen if `ExecRestoreHook` is not-empty.